Fpm

kzidane opened the pull request:
@dmalan and @glennholloway, how does that look?

glennholloway commented:
Very nice job!

When I run

    sudo dpkg --force-confnew --force-confmiss -i ide50_94_amd64.deb

in a fresh (version 69) workspace, I get:

    Unpacking ide50 (94) over (69) ...
    dpkg: warning: unable to delete old directory '/home/ubuntu/bin': Directory not empty

and then the instruction to run `update50` again. Is there a way to simplify that?

kzidane commented:
@glennholloway, guess no harm from that? I think the reason is that I'm not letting `ide50.deb` [overwrite](https://github.com/cs50/ide50/blob/master/Makefile#L24-L25) `debug50` anymore, and since `/home/ubuntu/bin` is missing from the new version, `dpkg` tries to delete it, but it aborts because the directory has other files.

It feels reasonable to do so as we may apply changes to our debug plugin that are not backward-compatible, and recall an older version may still be deployed online, which could cause problems?

Alternatively, perhaps we could add a `sed` to reset `"project/cs50/debug/@ver"`, letting the debug plugin itself take care of [overwriting](https://github.com/cs50/harvard.cs50.debug/blob/master/debug.js#L288) `debug50` in the next reload?

glennholloway commented:
@kzidane: So the reason that upgrading to release 94 will delete the `debug50` script is that it used to be listed among the package's files, but isn't so listed in release 94. Right?

Your suggestion for getting it back is that we reset `cs50` > `debug` > `@ver` to zero in `~/workspace/.c9/project.settings` and then ask the user to refresh her browser window after running `update50`? Is there no way for us to force the reload as part of `update50`?

BTW, do you understand why the new deb is 1/3 bigger than the old?

kzidane commented:
@glennholloway, 

> So the reason that upgrading to release 94 will delete the `debug50` script is that it used to be listed among the package's files, but isn't so listed in release 94. Right?

Yeah, it actually tries to delete the entire `/home/ubuntu/bin` as well, but it aborts because `.info50` lives there as well. In fact, just noticed it does delete `debug50` (thought it'd try to delete the entire directory, fail, and end up leaving `debug50` as is), so it sounds a must now to have the file rewritten.

> Your suggestion for getting it back is that we reset `cs50` > `debug` > `@ver` to zero in `~/workspace/.c9/project.settings` and then ask the user to refresh her browser window after running `update50`?

Yes, that's exactly right.

> Is there no way for us to force the reload as part of `update50`?

Hm, there may be a way. Will try a few things and let you know! There's a way without reloading, offline, but perhaps reloading will be better one and for all solution.

> BTW, do you understand why the new deb is 1/3 bigger than the old?

Not sure yet. Will let you know, if I could figure out why.

kzidane commented:
@glennholloway, how about avoiding reloads by setting a [watcher](https://apidoc.c9.io/c9v3/#!/api/watcher-event-delete) on `debug50` to have it written again, if it gets deleted? Do you see any problems with that?

latearthurdent commented:
@kzidane: the watcher would entail a PR to C9, no? So in order to switch over to your fpm method of package generation, we'd have to wait for that to be merged? (Or else warn students to reload the IDE page.)
It's a shame we can't find a way to tell `dpkg` to ignore the file that has been dropped since the last ide50 package update. It happily does exactly that for configuration files that have been omitted. Why not for regular files? (sigh)

glennholloway commented:
@kzidane: the watcher would entail a PR to C9, no? So in order to switch over to your fpm method of package generation, we'd have to wait for that to be merged? (Or else warn students to reload the IDE page.)
It's a shame we can't find a way to tell dpkg to ignore the file that has been dropped since the last ide50 package update. It happily does exactly that for configuration files that have been omitted. Why not for regular files? (sigh)

kzidane commented:
@glennholloway, they just merged a PR this morning. Guess we could open another, so that both PR's end up getting deployed at once, hopefully (typically takes a couple of days).

glennholloway commented:
@kzidane, so you'd construct a watcher as an addon to the debug plugin, such that when `delete` fires, the debug plugin would be reloaded?

kzidane commented:
@glennholloway, nope. Just rewrite `debug50`. No need for the whole plugin to be reloaded (if that's what you mean).

glennholloway commented:
@kzidane, But the content of the `debug50` script would come from the plugin, right? Not a separate copy of the file's text.

kzidane commented:
@glennholloway, yes, exactly. The script that is associated with the deployed version of the plugin.

glennholloway commented:
@kzidane, Sounds good.

kzidane commented:
@glennholloway, cs50/harvard.cs50.debug#8 has been deployed. Just rebased this and fixed the conflicts. Should be ready to merge, unless you have any takes!

dmalan commented:
Awesome, thanks so much. Let me defer to @glennholloway for final call!

dmalan commented on Makefile:
```diff
@@ -1,47 +1,68 @@
+	-t deb \
+	-v $(VERSION) \
+	--vendor CS50 \
+	--depends apache2 \
```
How come this one's a `depends` but the others are `deb-recommends`?

glennholloway replied:
@kzidane, @dmalan: No, I was also puzzled about the `apache2` exception.

glennholloway commented:
@kzidane,

Thanks for all this good work!

Does `update50` still need to replace `python2` with `python3` in project.settings? When will we be able to take that out?

Also, why does `project.settings` list `/usr/local/lib/python2.7/dist-packages` in `python/@path` (and in fact, puts it first)?

Did you ever get an inkling about why the .deb file is so much bigger when generated via FPM?

kzidane commented:
@glennholloway,

> Does update50 still need to replace python2 with python3 in project.settings? When will we be able to take that out?

Dropped per 1e02c2b. Thanks!

> Also, why does project.settings list /usr/local/lib/python2.7/dist-packages in python/@path (and in fact, puts it first)?

Hm, this should definitely not be the case. Do you have `project/cs50/simple/python/@configured` set in your `project.settings` as well? (Not a cs50/ide50 issue though)

> Did you ever get an inkling about why the .deb file is so much bigger when generated via FPM?

Searched, but afraid haven't managed to find out why yet!

glennholloway commented:
@kzidane, Yes the `cs50/simple/python/@configured` flag is true in `project.settings`, though I'm not sure what it means.

kzidane commented:
@dmalan, emailed!

@glennholloway, the way this works is that our simple plugin checks whether `project/cs50/simple/python/@configured` is set, in which case if it's not, it sets it, then sets all the pylint-related settings in the `project.settings`.

After the plugins sets these settings, if the user then changes these settings manually, this prevents the plugin from changing these settings back in the next reload.

What happens if you remove the `project/cs50/simple/python/@configured` setting or set it to `false`, then reload? Does `pythonpath` stays the same?
